File: amazon_clone/server/models/notification.js
const mongoose = require('mongoose');

const notificationSchema = new mongoose.Schema({
  userId: {
    type: String,
    required: true,
  },
  title: {
    type: String,
    required: true,
  },
  message: {
    type: String,
    required: true,
  },
  type: {
    type: String,
    required: true,
    enum: [
      'order_cancel_request', 
      'order_cancel_approved', 
      'order_cancel_rejected', 
      'order_update',
      'add_to_cart',
      'payment_success',
      'order_status_update'
    ],
  },
  data: {
    type: mongoose.Schema.Types.Mixed,
    default: null,
  },
  isRead: {
    type: Boolean,
    default: false,
  },
  createdAt: {
    type: Number,
    default: () => new Date().getTime(),
  },
});

const Notification = mongoose.model('Notification', notificationSchema);
module.exports = Notification;

File: amazon_clone/lib/models/notification.dart
import 'dart:convert';

class Notification {
  final String id;
  final String userId;
  final String title;
  final String message;
  final String
  type; // 'order_cancel_request', 'order_cancel_approved', 'order_cancel_rejected', 'add_to_cart', 'payment_success', 'order_status_update'
  final Map<String, dynamic>? data; // Additional data (orderId, etc.)
  final bool isRead;
  final int createdAt;

  Notification({
    required this.id,
    required this.userId,
    required this.title,
    required this.message,
    required this.type,
    this.data,
    this.isRead = false,
    required this.createdAt,
  });

  Map<String, dynamic> toMap() {
    return {
      'id': id,
      'userId': userId,
      'title': title,
      'message': message,
      'type': type,
      'data': data,
      'isRead': isRead,
      'createdAt': createdAt,
    };
  }

  factory Notification.fromMap(Map<String, dynamic> map) {
    return Notification(
      id: map['_id'] ?? '',
      userId: map['userId'] ?? '',
      title: map['title'] ?? '',
      message: map['message'] ?? '',
      type: map['type'] ?? '',
      data: map['data'],
      isRead: map['isRead'] ?? false,
      createdAt: map['createdAt']?.toInt() ?? 0,
    );
  }

  String toJson() => json.encode(toMap());

  factory Notification.fromJson(String source) =>
      Notification.fromMap(json.decode(source));

  Notification copyWith({
    String? id,
    String? userId,
    String? title,
    String? message,
    String? type,
    Map<String, dynamic>? data,
    bool? isRead,
    int? createdAt,
  }) {
    return Notification(
      id: id ?? this.id,
      userId: userId ?? this.userId,
      title: title ?? this.title,
      message: message ?? this.message,
      type: type ?? this.type,
      data: data ?? this.data,
      isRead: isRead ?? this.isRead,
      createdAt: createdAt ?? this.createdAt,
    );
  }
}
